### PR TITLE
Refactor indexers configuration section to add support for multiple indexers

### PR DESCRIPTION
**Pull request by rsevilla87**
## Type of change

- [x] Refactor
- [x] New feature
- [ ] Bug fix
- [x] Optimization
- [ ] Documentation Update

## Description

Adding a lot of interesting optimizations to the way this part of the configuration is handled. It will help to close https://github.com/kube-burner/kube-burner-ocp/issues/15, after this PR more changes are still needed to close it.

Relevant changes:

- Removal of _IndexerConfig_ from _globalConfig_. ⚠️   No backwards compatible
- New _indexers_ field at the same level of _globalConfig_, which enumerates the configured indexers. ⚠️   No backwards compatible
- Update `Run()` from from.go to accept metricsScraper as argument to reduce the number of arguments it receives
- When indexing happens, for each metric/alert/datapoint kube-burner iterates over the list of indexers
- ScraperConfig now accepts a list of metricsProfiles and alertProfiles, this change has been done on preparation for future changes to allow associating metricsProfiles and alertProfiles to specific indexers

New syntax now looks like:

```yaml
global:
  gc: true
  measurements:
    - name: podLatency
 indexers:                                       
   - type: opensearch                            
     esServers: ["blablabla:9200"]             
     defaultIndex: indexName   
   - type: local                                 
     metricsDirectory: collected-metrics
```

## Related Tickets & Documents

- Related Issue https://github.com/kube-burner/kube-burner-ocp/issues/15
- Closes 

cc @afcollins @jtaleric

### CODE REVIEW

File: pkg/util/metrics/metrics.go [hunk 1]
```diff
@@ -23,45 +23,61 @@ import (
...
-				Endpoint:     metricsScraperConfig.URL,
-				Token:        metricsScraperConfig.Token,
-				Profile:      metricsScraperConfig.MetricsProfile,
-				AlertProfile: metricsScraperConfig.AlertProfile,
-				Username:     metricsScraperConfig.Username,
-				Password:     metricsScraperConfig.Password,
+	if scraperConfig.MetricsEndpoint != "" {
+		DecodeMetricsEndpoint(scraperConfig.MetricsEndpoint, &metricsEndpoints)
+	} else {
+		for _, metricsProfile := range scraperConfig.MetricsProfiles {
```
Comment (rsevilla87) [inline pkg/util/metrics/metrics.go:63]: Create a new metricsEndpoint for each metricProfile and alertProfile

File: pkg/workloads/helpers.go [hunk 2]
```diff
@@ -105,79 +101,19 @@ func (wh *WorkloadHelper) Run(workload string, metricsProfiles []string, alertsP
...
-				alertM, err := alerting.NewAlertManager(alertsProfile, wh.Metadata.UUID, indexer, p, embedConfig)
-				if err != nil {
-					log.Fatal(err)
-				}
-				alertMs = append(alertMs, alertM)
-			}
-		}
-	}
-	rc, err = burner.Run(ConfigSpec, prometheusClients, alertMs, indexer, wh.Timeout, wh.MetricsMetadata)
+	metricsScraper = metrics.ProcessMetricsScraperConfig(metrics.ScraperConfig{
```
Comment (rsevilla87) [inline pkg/workloads/helpers.go:104]: All the logic above is actually performed by ProcessMetricsScraperConfig()

File: test/k8s/metrics-b0f82842-aa6f-4881-9e53-0a3d0ff59f09/jobSummary-namespaced.json [hunk 3]
```diff
@@ -0,0 +1 @@
+[{"timestamp":"2024-02-29T21:05:36.695849854Z","endTimestamp":"2024-02-29T21:08:04.529749537Z","elapsedTime":148,"uuid":"b0f82842-aa6f-4881-9e53-0a3d0ff59f09","metricName":"jobSummary","jobConfig":{"jobIterations":5,"jobIterationDelay":5000000000,"name":"namespaced","jobType":"create","qps":3,"burst":3,"namespace":"namespaced","maxWaitTimeout":14400000000000,"waitForDeletion":true,"waitWhenFinished":true,"cleanup":true,"namespacedIterations":true,"iterationsPerNamespace":1,"verifyObjects":true,"errorOnVerify":true,"preLoadImages":true,"preLoadPeriod":5000000000,"churn":true,"churnCycles":2,"churnPercent":20,"churnDuration":60000000000,"churnDelay":5000000000,"churnDeletionStrategy":"default"},"metadata":{},"version":"multiple-indexers@1ee23893802664db889f3bb1ca096d0bb1ccd4ab"}]
```
Comment (vishnuchalla) [inline test/k8s/metrics-b0f82842-aa6f-4881-9e53-0a3d0ff59f09/jobSummary-namespaced.json:1]: Do we need this test metrics here>

Comment (rsevilla87) [reply test/k8s/metrics-b0f82842-aa6f-4881-9e53-0a3d0ff59f09/jobSummary-namespaced.json:1]: my bad I committed it by mistake

File: pkg/burner/job.go [hunk 4]
```diff
@@ -192,22 +194,24 @@ func Run(configSpec config.Spec, prometheusClients []*prometheus.Prometheus, ale
...
-			if indexer != nil && !job.SkipIndexing {
-				msWg.Add(1)
-				go func(jobName string) {
-					defer msWg.Done()
-					measurements.Index(indexer, jobName)
-				}(job.Name)
+			if !job.SkipIndexing {
+				for _, indexer := range metricsScraper.IndexerList {
+					msWg.Add(1)
+					go func(jobName string, indexer indexers.Indexer) {
```
[the remark this answers is not in the document]
Comment (rsevilla87) [reply pkg/burner/job.go:210]: This goroutines index the documents of the measurements, as they don't depend on Prometheus we can spawn them in their own goroutine without taking the risk of overloading Prometheus.

File: docs/reference/configuration.md [hunk 5]
```diff
@@ -2,12 +2,19 @@
...
 
-It is possible to use [go-template](https://pkg.go.dev/text/template) semantics within this configuration file. It is also important to note that every environment variable is passed to this template, so we can reference them using the syntax `{{.MY_ENV_VAR}}`. For example, you could define the `indexerConfig` section of your own configuration file, such as:
+It is possible to use [go-template](https://pkg.go.dev/text/template) semantics within this configuration file. It is also important to note that every environment variable is passed to this template, so we can reference them using the syntax `{{.MY_ENV_VAR}}`. For example, you could define the `indexers` section of your own configuration file, such as:
 
 ```yaml
-type: elastic
-esServers: [{{ .ES_SERVER }}]
-defaultIndex: elasticsearch-index
+indexers:
+{{ if .OS_INDEXING }}
```
[the remark this answers is not in the document]
Comment (rsevilla87) [reply docs/reference/configuration.md:9]: Renamed

**Review by vishnuchalla — APPROVED** [review]
Assuming that more features to come along with upcoming PRs as mentioned in the description, good with shipping it.

**Pull request merged**